README.md
# Vedran

> Polkadot chain load balancer.

### Architecture

_Vedran loadbalancer_ is used in conjunction with [Vedran daemon](https://github.com/NodeFactoryIo/vedran-daemon).
Suppose the node owner wants to register to loadbalancer, than it is required to install and run _Vedran daemon_.
Daemon executes the registration process and starts providing all relevant information (ping, metrics) to the _Vedran loadbalancer_.
Please check [Vedran daemon repo](https://github.com/NodeFactoryIo/vedran-daemon) for more details on the daemon itself.

![Image of vedran architecture](./assets/vedran-arch.png)

## Demo

_This is dockerized demo of entire setup with loadbalancer, node and daemon_

### Requirements

- Install [Docker Engine](https://docs.docker.com/engine/install/)
- Install [Docker Compose](https://docs.docker.com/compose/install/)

**Run demo with `docker-compose up`**

_After all components have been started and node has sent first valid metrics report (after 30 seconds),
you can invoke RPC methods on `localhost:4000` using HTTP requests or on `localhost:4000/ws` using WebSocket request_
Metrics can be seen on `localhost:3000` hosted grafana under `vedran-dashboard`

You can check available rpc methods with:

```bash
curl -H "Content-Type: application/json" -d '{"id":1, "jsonrpc":"2.0", "method": "rpc_methods"}' http://localhost:4000
```
This demo starts five separate dockerized components:
- _Polkadot node_ ([repository](https://github.com/paritytech/polkadot))
- _Vedran daemon_ ([repository](https://github.com/NodeFactoryIo/vedran-daemon))
- _Vedran loadbalancer_ (port: 4000)
- _Prometheus server_ (port: 9090) - scrapes metrics from vedran's `/metrics` endpoint
- _Grafana_ (port: 3000) - Visualizes metrics [Example dashboard](http://localhost:3000/d/7tNmHxAGz/vedran-dashboard?orgId=1&refresh=5s)


### Trigger Manual Payout

Our compose setup runs dev chain, and our load balancer uses Allice account to do payout, 
so you don't have to obtain dev DOTs. Polkadot node operator is Bob (he received payout from Allice).
Load balancer in this setup runs payout daily, if you don't want to wait,
you can run following command which will create an additional container (in compose network) and trigger payout from Allice account:

```
docker run --network vedran_default nodefactory/vedran:latest payout --private-key 0xe5be9a5092b81bca64be81d212e7f2f9eba183bb7a90954f7b76361f6edb5c0a --payout-reward 100 --load-balancer-url "http://vedran:4000/ws"
```

## Get **vedran** binary releases

Download prebuild binary from [releases](https://github.com/NodeFactoryIo/vedran/releases).
Be careful to chose appropriate binary depending on your OS. For more details on how to run _vedran loadbalancer_ see [Starting loadbalancer](#starting-loadbalancer) part.

## Get **vedran** package
Alternatively, it is possible to get _vedran_ golang package:
1. Install [Golang](https://golang.org/doc/install) **1.15 or greater**
2. Run the command below
```
go get github.com/NodeFactoryIo/vedran
```
3. Run vedran from your Go bin directory. For linux systems it will likely be:
```
~/go/bin/vedran
```
Note that if you need to do this, you probably want to add your Go bin directory to your $PATH to make things easier!


## Starting loadbalancer

First download latest prebuilt binaries [from releases](https://github.com/NodeFactoryIo/vedran/releases) and unzip it.

Load balancer is started by invoking **start** command.

For example `./vedran start --auth-secret=supersecret --private-key=lb-wallet-private-key`.

For more information you can always run vedran with `--help` flag.
For list of all commands run `vedran --help` or for list of all options for specific command run `vedran start --help`.

**Load balancer will expose Polkadot RPC API on port 80 by default (can be changed using flag `--server-port`)**

Vedran loadbalancer supports both **HTTP** and **Websockets** protocols for Polkadot RPC API.

- **HTTP - available on root path** `/`
- **WS - available on separate path** `/ws`

**For production use certificates (e.g. https://certbot.eff.org/) should be generated and passsed via flags: `--key-file`, `--cert-file` and port changed to 443**

Start command will start application on 2 ports that need to be exposed to public:
 1. RPC entrypoint to nodes and API for nodes to register to load balancer (default: 80)
 2. HTTP tunnel server for creating tunnels between the node and load balancer so node operators don't to have expose nodes to public network (default: 5223)

### Required flags

`--auth-secret` - authentication secret used for generating tokens

`--private-key` - loadbalancers wallet private key, used for sending founds on payout

### Most important flags

| Flag | Description | Default value |
|----|-----------|:--------:|
|`--server-port`|port on which RPC API is exposed|80|
|`--public-ip`|public IP address of loadbalancer|uses multiple services to find out public IP|
|`--cert-file`|path to SSL certification file|uses HTTP|
|`--key-file`|path to SSL private key file|uses HTTP|
|`--tunnel-port`|port on which tunnel server is listening for connect requests|5223|
|`--tunnel-port-range`|range of ports that will be used for creating tunnels|20000:30000|

### Other flags

| Flag | Description | Default value |
|----|-----------|:--------:|
|`--name`|public name for load balancer|autogenerated name is used|
|`--capacity`|maximum number of nodes allowed to connect|unlimited capacity|
|`--whitelist`|comma separated list of node id-s, if provided only these nodes will be allowed to connect. This flag can't be used together with --whitelist-file flag, only one option for setting whitelisted nodes can be used|all nodes are whitelisted|
|`--whitelist-file`|path to file with node id-s in each line, if provided only these nodes will be allowed to connect. This flag can't be used together with --whitelist flag, only one option for setting whitelisted nodes can be used|all nodes are whitelisted|
|`--fee`|value between 0-1 representing fixed fee percentage that loadbalancer will take|0.1 (10%)|
|`--selection`|type of selection that is used for selecting nodes on new request, valid values are `round-robin` and `random`|`round-robin`|
|`--payout-interval`|automatic payout interval specified as number of days, for more details see [payout instructions](#payouts)|-|
|`--payout-reward`|defined reward amount that will be distributed on the payout (amount in Planck), for more details see [payout instructions](#payouts)|-|
|`--lb-payout-address`|address on which load balancer fee will be sent|-|
|`--log-level`|log level (debug, info, warn, error)|error|
|`--log-file`|path to file in which logs will be saved|`stdout`|
|`--root-dir`|root directory for all generated files (e.g. database file, log file)|uses current directory|

### Obtaining DOTs
If you want to do anything on Polkadot, Kusama, or Westend, then you'll need to get an account and some DOT, KSM, or WND tokens, respectively.
When initializing payout, you will provide loadbalancer with created account and from this account rewards will be sent to connected nodes on payout.

For Westend's WND tokens, see the faucet [instructions on the Wiki](https://wiki.polkadot.network/docs/en/learn-DOT#getting-westies).

## Payouts

### Automatic payout

When starting _vedran loadbalancer_ it is possible to configure automatic payout by providing these flags:

`--private-key` - loadbalancers wallet private key (string representation of hex value prefixed with 0x), used for sending rewards on the payout

`--payout-interval` - automatic payout interval specified as number of days

`--payout-reward` - defined total reward amount that will be distributed on the payout (amount in Planck). If omitted, the entire balance of lb wallet will be used as a total reward, and in this case `--lb-payout-fee-address` must be set

`--lb-payout-address` - address on which load balancer fee will be sent. If omitted, load balancer fee will be left on load balancer wallet after payout. This flag is **required** if `--payout-reward` is not set (or set to -1)

If all flags have been provided, then each {_payout-interval_} days automatic payout will be started.

### Manual payout

It is possible to run payout script at any time by invoking `vedran payout` command through the console.

`--private-key` - loadbalancers wallet private key (string representation of hex value prefixed with 0x), used for sending rewards on the payout

`--payout-reward` - defined total reward amount that will be distributed on the payout (amount in Planck). If omitted, the entire balance of lb wallet will be used as a total reward, and in this case `--lb-payout-fee-address` must be set

`--lb-payout-fee-address` - address on which load balancer fee will be sent. If omitted, load balancer fee will be left on load balancer wallet after payout. This flag is **required** if `--payout-reward` is not set (or set to -1)

Additionally, it is possible to change URL on which payout script will connect with loadbalancer when executing transactions by setting flag (default value will be _http://localhost:80_)

`--load-balancer-url` - loadbalancer URL

### Get private key
You can use [subkey](https://substrate.dev/docs/en/knowledgebase/integrate/subkey) tool to get private key for your wallet.

After installing subkey tool call `subkey inspect "insert your mnemonic here"`.
You can find private key as _Secreet seed_. See example output of subkey command:

```
  Secret seed:      0x1a84771145cdcee05e49142aaff2e5d669ce4b29344a09b973b751ae661acabf
  Public key (hex): 0xa4548fa9b3b15dc4d1c59789952f0ccf6138dd63faf802637895c941f0522d35
  Account ID:       0xa4548fa9b3b15dc4d1c59789952f0ccf6138dd63faf802637895c941f0522d35
  SS58 Address:     5FnAq6wrMzri5V6jLfKgBkbR2rSAMkVAHVYWa3eU7TAV5rv9
```

## Monitoring

Monitoring is done via grafana and prometheus which are expected to be installed.

### Installation
 - [Grafana installation](https://grafana.com/docs/grafana/latest/installation/)
 - [Prometheus installation](https://prometheus.io/docs/prometheus/latest/installation/)

### Configuration

-  ### Grafana
    Should be configured to fetch data from prometheus server as data source [Tutorial](https://prometheus.io/docs/visualization/grafana/).

    Should have a dashboard that visualizes data scraped from prometheus server. Example configuration can be found [here](./infra/grafana/provisioning/dashboards/vedran-dashboard.json) and can be imported like [this](https://grafana.com/docs/grafana/latest/dashboards/export-import/#importing-a-dashboard).


-  ### Prometheus
    Prometheus should be configured to scrape metrics from vedran's `/metrics` endpoint
    via prometheus .yml configuration. Example of which can be found [here](./infra/prometheus/prometheus.yml)
   
### Stats collection intervals

It is possible to change default calculation intervals for specific statistic categories. These intervals define how
often will these statistics be recalculated.

Valid values are time intervals such as "5s", "1.5h" or "2h45m". Valid time units are "ms", "s", "m", "h"

| ENV | Description | Default value |
|----|-----------|:--------:|
|`PROM_FEE_STATS_INTERVAL`|payout fees for nodes and load balancer|12 hours|
|`PROM_NODE_STATS_INTERVAL`|active and penalized nodes|15 seconds|
|`PROM_REQUEST_STATS_INTERVAL`|successful and failed requests|15 seconds|
|`PROM_PAYOUT_STATS_INTERVAL`|payout distribution|1 minute|


## Vedran loadbalancer API

`POST   api/v1/nodes`

Register node to loadbalancer. Body should contain details about node:

```json
{
  "id": "string",
  "config_hash": "string",
  "payout_address": "string"
}
```

Returns **token** used for invoking rest of API and **tunnel_server_address** on which daemon can open tunnel toward loadbalancer.

```json
{
  "token": "string",
  "tunnel_server_address": "string"
}
```

---

`POST   api/v1/nodes/pings`

Ping loadbalancer from node. Auth token should be in header as `X-Auth-Header`.

---

`PUT    api/v1/nodes/metrics`

Send metrics for node. Auth token should be in header as `X-Auth-Header`. Body should contain metrics as:

```json
{
  "peer_count": "int32",
  "best_block_height": "int64",
  "finalized_block_height": "int64",
  "target_block_height": "int64",
  "ready_transaction_count": "int32"
}
```

---

`GET    api/v1/stats`

Returns statistics for all nodes (mapped on node payout address).

```json
{
  "stats": {
    "node_1_payout_address": {
      "total_pings": "float64",
      "total_requests": "float64"
    },
    "node_2_payout_address": {
      "total_pings": "float64",
      "total_requests": "float64"
    }
  }
}
```

---

`GET    api/v1/stats/lb`

Returns statistics on reward distribution between load balancer and nodes.

```json
{
  "lb_fee": "string",
  "nodes_fee": "string"
}
```

## Development

### Clone

```bash
git clone git@github.com:NodeFactoryIo/vedran.git
```

### Lint
[Golangci-lint](https://golangci-lint.run/usage/install/#local-installation) is expected to be installed.

```bash
make lint
```

### Build

```bash
make build
```

### Test

```bash
make test
```

## License

This project is licensed under Apache 2.0:
- Apache License, Version 2.0, ([LICENSE-APACHE](http://www.apache.org/licenses/LICENSE-2.0))


CHANGELOG.md
# Changelog

## [unreleased]((https://github.com/NodeFactoryIo/vedran/tree/HEAD))
[Full Changelog](https://github.com/NodeFactoryIo/vedran/compare/v0.5.5...HEAD)

### Added

### Fix
- Fix duplicated ports inside tunnel address pool [\#204](https://github.com/NodeFactoryIo/vedran/pull/204) ([mpetrun5](https://github.com/mpetrun5))

### Changed

## [v0.5.5]((https://github.com/NodeFactoryIo/vedran/tree/v0.5.5))
[Full Changelog](https://github.com/NodeFactoryIo/vedran/compare/v0.5.4...v0.5.5)

### Added

### Fix
- Fix request counters in grafana dashboard [\#200](https://github.com/NodeFactoryIo/vedran/pull/200) ([mpetrun5](https://github.com/mpetrun5))

### Changed

## [v0.5.4]((https://github.com/NodeFactoryIo/vedran/tree/v0.5.4))
[Full Changelog](https://github.com/NodeFactoryIo/vedran/compare/v0.5.3...v0.5.4)

### Added

### Fix
- Fix tunnel tcp connections not closing after requests finish [\#197](https://github.com/NodeFactoryIo/vedran/pull/197) ([mpetrun5](https://github.com/mpetrun5))
- Ping handling upgrade [\#198](https://github.com/NodeFactoryIo/vedran/pull/198) ([MakMuftic](https://github.com/MakMuftic))

### Changed

## [v0.5.3]((https://github.com/NodeFactoryIo/vedran/tree/v0.5.3))
[Full Changelog](https://github.com/NodeFactoryIo/vedran/compare/v0.5.2...v0.5.3)

### Added

### Fix
- Minor bug fixes [\#194](https://github.com/NodeFactoryIo/vedran/pull/194) ([MakMuftic](https://github.com/MakMuftic))

### Changed
- Upgrade request display inside grafana dashboard [\#190](https://github.com/NodeFactoryIo/vedran/pull/190) ([MakMuftic](https://github.com/MakMuftic))

## [v0.5.2]((https://github.com/NodeFactoryIo/vedran/tree/v0.5.2))
[Full Changelog](https://github.com/NodeFactoryIo/vedran/compare/v0.5.1...v0.5.2)

### Added
- Add additional logging [\#188](https://github.com/NodeFactoryIo/vedran/pull/188) ([MakMuftic](https://github.com/MakMuftic))

### Fix

### Changed

## [v0.5.1]((https://github.com/NodeFactoryIo/vedran/tree/v0.5.1))
[Full Changelog](https://github.com/NodeFactoryIo/vedran/compare/v0.5.0...v0.5.1)

### Added

### Fix
- Add buffer for target block metric [\#184](https://github.com/NodeFactoryIo/vedran/pull/184) ([MakMuftic](https://github.com/MakMuftic))

### Changed

## [v0.5.0]((https://github.com/NodeFactoryIo/vedran/tree/v0.5.0))
[Full Changelog](https://github.com/NodeFactoryIo/vedran/compare/v0.4.3...v0.5.0)

### Added
- Add target block metric [\#177](https://github.com/NodeFactoryIo/vedran/pull/177) ([MakMuftic](https://github.com/MakMuftic))

### Fix
- Fix min TLS version [\#175](https://github.com/NodeFactoryIo/vedran/pull/175) ([MakMuftic](https://github.com/MakMuftic))

### Changed
- Replace go-substrate-rpc-client fork [\#176](https://github.com/NodeFactoryIo/vedran/pull/176) ([MakMuftic](https://github.com/MakMuftic))

## [v0.4.3]((https://github.com/NodeFactoryIo/vedran/tree/v0.4.3))
[Full Changelog](https://github.com/NodeFactoryIo/vedran/compare/v0.4.2...v0.4.3\)

### Added
- Add CLI flag for root dir [\#172](https://github.com/NodeFactoryIo/vedran/pull/172) ([MakMuftic](https://github.com/MakMuftic))

### Fix

### Changed

## [v0.4.2]((https://github.com/NodeFactoryIo/vedran/tree/v0.4.2))
[Full Changelog](https://github.com/NodeFactoryIo/vedran/compare/v0.4.1...v0.4.2\)

### Added

### Fix

### Changed
- Custom stats collecting intervals [\#168](https://github.com/NodeFactoryIo/vedran/pull/168) ([MakMuftic](https://github.com/MakMuftic))

## [v0.4.1]((https://github.com/NodeFactoryIo/vedran/tree/v0.4.1))
[Full Changelog](https://github.com/NodeFactoryIo/vedran/compare/v0.4.0...v0.4.1\)

### Added

### Fix
- Fix manual payout [\#165](https://github.com/NodeFactoryIo/vedran/pull/165) ([MakMuftic](https://github.com/MakMuftic))

### Changed

## [v0.4.0]((https://github.com/NodeFactoryIo/vedran/tree/v0.4.0))
[Full Changelog](https://github.com/NodeFactoryIo/vedran/compare/v0.3.1...v0.4.0)

### Added
- Provide LB fee information [\#158](https://github.com/NodeFactoryIo/vedran/pull/158) ([MakMuftic](https://github.com/MakMuftic))

### Fix
- Metrics endpoint and grafana dashboard example [\#138](https://github.com/NodeFactoryIo/vedran/pull/138) ([mpetrun5](https://github.com/mpetrun5))
- Fix next payout date prometheus value [\#161](https://github.com/NodeFactoryIo/vedran/pull/161) ([mpetrun5](https://github.com/mpetrun5))

### Changed

## [v0.3.1]((https://github.com/NodeFactoryIo/vedran/tree/v0.3.1))
[Full Changelog](https://github.com/NodeFactoryIo/vedran/compare/v0.3.0...v0.3.1)

### Added

### Fix
- Enable CORS and set upgrader CheckOrigin to true  [\#149](https://github.com/NodeFactoryIo/vedran/pull/149) ([mpetrun5](https://github.com/mpetrun5))

### Changed

## [v0.3.0]((https://github.com/NodeFactoryIo/vedran/tree/v0.3.0))

[Full Changelog](https://github.com/NodeFactoryIo/vedran/compare/v0.2.0...v0.3.0)

### Added
- Check if port range valid [\#104](https://github.com/NodeFactoryIo/vedran/pull/104) ([mpetrun5](https://github.com/mpetrun5))
- Valid flag on node [\#105](https://github.com/NodeFactoryIo/vedran/pull/105) ([mpetrun5](https://github.com/mpetrun5))
- Passing SSL certificates [\#112](https://github.com/NodeFactoryIo/vedran/pull/112) ([mpetrun5](https://github.com/mpetrun5))
- Expose stats endpoints [\#114](https://github.com/NodeFactoryIo/vedran/pull/114) ([MakMuftic](https://github.com/MakMuftic))
- Calculating reward distribution [\#124](https://github.com/NodeFactoryIo/vedran/pull/124) ([MakMuftic](https://github.com/MakMuftic))
- Add payout CLI command [\#126](https://github.com/NodeFactoryIo/vedran/pull/126) ([MakMuftic](https://github.com/MakMuftic))
- Support WS connections [#\132](https://github.com/NodeFactoryIo/vedran/pull/132) ([MakMuftic](https://github.com/MakMuftic))
- Execute payout transactions [\#127](https://github.com/NodeFactoryIo/vedran/pull/127) ([MakMuftic](https://github.com/MakMuftic))
- Sign stats request [\#143](https://github.com/NodeFactoryIo/vedran/pull/143) ([MakMuftic](https://github.com/MakMuftic))

cmd/start.go
package cmd

import (
	"errors"
	"fmt"
	"net/url"
	"os"
	"path"
	"strconv"
	"strings"

	"github.com/NodeFactoryIo/vedran/internal/whitelist"

	"github.com/NodeFactoryIo/vedran/internal/configuration"
	"github.com/NodeFactoryIo/vedran/internal/ip"
	"github.com/NodeFactoryIo/vedran/internal/loadbalancer"
	"github.com/NodeFactoryIo/vedran/internal/tunnel"
	"github.com/NodeFactoryIo/vedran/pkg/http-tunnel/server"
	"github.com/NodeFactoryIo/vedran/pkg/logger"
	"github.com/NodeFactoryIo/vedran/pkg/util"
	"github.com/NodeFactoryIo/vedran/pkg/util/random"
	log "github.com/sirupsen/logrus"
	"github.com/spf13/cobra"
)

var (
	// load balancer related flags
	authSecret     string
	name           string
	certFile       string
	keyFile        string
	capacity       int64
	whitelistArray []string
	whitelistFile  string
	fee            float32
	selection      string
	serverPort     int32
	publicIP       string
	rootDir        string
	// payout related flags
	payoutFeeAddress           string
	payoutPrivateKey           string
	payoutNumberOfDays         int32
	payoutTotalReward          string
	payoutTotalRewardAsFloat64 float64
	autoPayoutDisabled         bool
	// logging related flags
	logLevel string
	logFile  string
	// tunnel related flags
	tunnelServerPort string
	tunnelPortRange  string
)

var startCmd = &cobra.Command{
	Use:   "start",
	Short: "Starts vedran load balancer",
	Run:   startCommand,
	PreRunE: func(cmd *cobra.Command, args []string) error {
		DisplayBanner()
		// create root dir
		if _, err := os.Stat(rootDir); os.IsNotExist(err) && rootDir != "" {
			err = os.Mkdir(rootDir, 0700)
			if err != nil {
				log.Fatal("Unable to create root dir ", err)
			}
		}
		// setup logger
		level, err := log.ParseLevel(logLevel)
		if err != nil {
			log.Fatalf("Invalid log level %s", logLevel)
		}
		logFilePath := ""
		if rootDir != "" && logFile != "" {
			logFilePath = path.Join(rootDir, logFile)
			fmt.Printf("Log file set to: %s", logFilePath)
		}
		err = logger.SetupLogger(level, logFilePath)
		if err != nil {
			return err
		}
		return nil
	},
	Args: func(cmd *cobra.Command, args []string) error {
		// valid values are round-robin and random
		if selection != "round-robin" && selection != "random" {
			return errors.New("invalid selection option selected")
		}
		// all positive integers are valid, and -1 representing unlimited capacity
		if capacity < -1 {
			return errors.New("invalid capacity value")
		}
		// valid value is between 0-1
		if fee < 0 || fee > 1 {
			return errors.New("invalid fee value")
		}
		// well known ports and registered ports
		if !util.IsValidPortAsInt(serverPort) {
			return errors.New("invalid rpc server port number")
		}
		// valid format is PortMin:PortMax
		prt := strings.Split(tunnelPortRange, ":")
		if len(prt) != 2 {
			return errors.New("invalid port range, should be defined as \"PortMin:PortMax\"")
		}
		if !util.IsValidPortAsStr(prt[0]) {
			return errors.New("invalid port number provided for min port inside port range")
		}
		if !util.IsValidPortAsStr(prt[1]) {
			return errors.New("invalid port number provided for max port inside port range")
		}

		// valid certificates
		if (certFile != "" && keyFile == "") || (keyFile != "" && certFile == "") {
			return errors.New("both cert and key file flags need to be set for valid certificate")
		}

		minPort, _ := strconv.Atoi(prt[0])
		maxPort, _ := strconv.Atoi(prt[1])
		if capacity == -1 {
			capacity = int64(maxPort - minPort)
		} else if int64(maxPort-minPort) < capacity {
			return errors.New("port range too small for target capacity")
		}

		if whitelistArray != nil && whitelistFile != "" {
			return errors.New("only one flag for setting whitelisted nodes should be set")
		}

		autoPayoutDisabled = payoutNumberOfDays == 0
		if !autoPayoutDisabled {
			if payoutNumberOfDays <= 0 {
				return errors.New("invalid payout interval")
			}
			rewardAsFloat64, err := ValidatePayoutFlags(payoutTotalReward, payoutFeeAddress, false)
			if err != nil {
				return err
			}
			payoutTotalRewardAsFloat64 = rewardAsFloat64
		}

		return nil
	},
}

func init() {
	startCmd.Flags().StringVar(
		&authSecret,
		"auth-secret",
		"",
		"[REQUIRED] Authentication secret used for generating tokens")

	startCmd.Flags().StringVar(
		&name,
		"name",
		fmt.Sprintf("load-balancer-%s", random.String(12, random.Alphabetic)),
		"[OPTIONAL] Public name for load balancer, autogenerated name used if omitted")

	startCmd.Flags().Int64Var(
		&capacity,
		"capacity",
		-1,
		"[OPTIONAL] Maximum number of nodes allowed to connect, where -1 represents no upper limit")

	startCmd.Flags().StringSliceVar(
		&whitelistArray,
		"whitelist",
		nil,
		"[OPTIONAL] Comma separated list of node id-s, if provided only these nodes will be allowed to connect."+
			"This flag can't be used together with --whitelist-file flag, only one option for setting whitelisted nodes can be used")

	startCmd.Flags().StringVar(
		&whitelistFile,
		"whitelist-file",
		"",
		"[OPTIONAL] Path to file with node id-s in each line that should be whitelisted."+
			"This flag can't be used together with --whitelist flag, only one option for setting whitelisted nodes can be used")

	startCmd.Flags().Float32Var(
		&fee,
		"fee",
		0.1,
		"[OPTIONAL] Value between 0-1 representing fee percentage")

	startCmd.Flags().StringVar(
		&selection,
		"selection",
		"round-robin",
		"[OPTIONAL] Type of selection used for choosing nodes (round-robin, random)")

	startCmd.Flags().StringVar(
		&certFile,
		"cert-file",
		"",
		"[OPTIONAL] SSL certificate file")

	startCmd.Flags().StringVar(
		&keyFile,
		"key-file",
		"",
		"[OPTIONAL] SSL matching private key")

	startCmd.Flags().Int32Var(
		&serverPort,
		"server-port",
		80,
		"[OPTIONAL] Port on which load balancer rpc server will be started")

	startCmd.Flags().StringVar(
		&publicIP,
		"public-ip",
		"",
		"[OPTIONAL] Public ip of load balancer")

	startCmd.Flags().StringVar(
		&logLevel,
		"log-level",
		"error",
		"[OPTIONAL] Level of logging (eg. info, warn, error)")

	startCmd.Flags().StringVar(
		&logFile,
		"log-file",
		"",
		"[OPTIONAL] Path to logfile (default stdout)")

	startCmd.Flags().StringVar(
		&tunnelServerPort,
		"tunnel-port",
		"5223",
		"[OPTIONAL] Address on which tunnel server is listening")

	startCmd.Flags().StringVar(
		&tunnelPortRange,
		"tunnel-port-range",
		"20000:30000",
		"[OPTIONAL] Range of ports which is used to open tunnels")

	startCmd.Flags().StringVar(
		&payoutPrivateKey,
		"private-key",
		"",
		"[REQUIRED] Load balancers wallet private key, used for sending funds on payout",
	)

	startCmd.Flags().StringVar(
		&payoutTotalReward,
		"payout-reward",
		"-1",
		"[OPTIONAL] Total reward pool in Planck. If omitted, total balance of load balancer wallet will be considered as payout reward",
	)

	startCmd.Flags().StringVar(
		&payoutFeeAddress,
		"lb-payout-address",
		"",
		"[OPTIONAL] Address on which load balancer fee will be sent. If omitted, load balancer fee will be left on load balancer wallet after payout")

	startCmd.Flags().Int32Var(
		&payoutNumberOfDays,
		"payout-interval",
		0,
		"[OPTIONAL] Payout interval in days, meaning each X days automatic payout will be executed")

	startCmd.Flags().StringVar(
		&rootDir,
		"root-dir",
		"",
		"[OPTIONAL] Root directory for all generated files (e.g. database file, log file)")

	_ = startCmd.MarkFlagRequired("private-key")

	RootCmd.AddCommand(startCmd)
}

func startCommand(_ *cobra.Command, _ []string) {
	// creating address pool
	pPool := &server.AddrPool{}
	err := pPool.Init(tunnelPortRange)
	if err != nil {
		log.Fatalf("Failed assigning port range because of: %v", err)
	}

	// defining tunnel server address
	var tunnelServerAddress string
	if publicIP == "" {
		IP, err := ip.Get()
		if err != nil {
			log.Fatal("Unable to fetch public IP address. Please set one explicitly!", err)
		}
		tunnelServerAddress = fmt.Sprintf("%s:%s", IP.String(), tunnelServerPort)
	} else {
		tunnelServerAddress = fmt.Sprintf("%s:%s", publicIP, tunnelServerPort)
	}
	log.Infof("Tunnel server will listen on %s and connect tunnels on port range %s", tunnelServerAddress, tunnelPortRange)

	// initializing whitelisting
	whitelistEnabled, err := whitelist.InitWhitelisting(whitelistArray, whitelistFile)
	if err != nil {
		log.Fatal("Unable to set whitelisted nodes ", err)
	}
	log.Debugf("Whitelisting set to: %t", whitelistEnabled)

	var payoutConfiguration *configuration.PayoutConfiguration
	if !autoPayoutDisabled {
		lbUrl, _ := url.Parse("http://" + publicIP + ":" + string(serverPort))
		payoutConfiguration = &configuration.PayoutConfiguration{
			PayoutNumberOfDays: int(payoutNumberOfDays),
			PayoutTotalReward:  payoutTotalRewardAsFloat64,
			LbFeeAddress:       payoutFeeAddress,
			LbURL:              lbUrl,
		}
	}

	tunnel.StartHttpTunnelServer(tunnelServerPort, pPool)
	loadbalancer.StartLoadBalancerServer(
		configuration.Configuration{
			AuthSecret:          authSecret,
			Name:                name,
			CertFile:            certFile,
			KeyFile:             keyFile,
			Capacity:            capacity,
			Fee:                 fee,
			Selection:           selection,
			Port:                serverPort,
			TunnelServerAddress: tunnelServerAddress,
			PortPool:            pPool,
			WhitelistEnabled:    whitelistEnabled,
			PayoutConfiguration: payoutConfiguration,
			RootDir:             rootDir,
		},
		payoutPrivateKey,
	)
}
